ARCHITECTURE.md

```diff
@@ -0,0 +1,196 @@
+  well as [TableAdmin](/google/cloud/bigtable/table_admin.h) and
+  [InstanceAdmin](/google/cloud/bigtable/instance_admin.h) clients.
+- Storage has a single [storage::Client](/google/cloud/storage/client.h) client
+  class.
+- Pub/Sub has two non-admin clients, the
+  [pubsub::Publisher](/google/cloud/pubsub/publisher.h)
+  and the [pubsub::Subscriber](/google/cloud/pubsub/subscriber.h). It also has
+  two admin clients:
+  [pubsub::TopicAdminClient](/google/cloud/pubsub/topic_admin_client.h),
+  and [pubsub::SubscriptionAdminClient](/google/cloud/pubsub/subscription_admin_client.h).
+- Spanner has one non-admin client, the
+  [spanner::Client](/google/cloud/spanner/client.h). It also has two admin
+  clients:
+  [pubsub::InstanceAdminClient](/google/cloud/spanner/instance_admin_client.h),
+  and
+  [pubsub::DatabaseAdminClient](/google/cloud/spanner/database_admin_client.h).
+
+Generally these classes are very "thin", they take function arguments from the
+application, package them in lightweight structure, and then forward the request
+to the `${library}::*Connection` class.
+
+It is important to know that *almost* always there is one RPC generated by each
+`*Client` member functions and RPCs. But that "almost" is very important, the
+devil is, as usual, in the details.
+
+## The `*Connection` classes
+
+Connections serve two functions:
+
+* They are interfaces, which applications can mock to implement tests against
+  the `google-cloud-cpp` library.
+* The default implementation injects the "retry loop" for each RPC.
+
+Most of the time there is a 1-1 mapping between a `FooClient` and the
+corresponding `FooConnection`.
+
+Because `*Connection` classes are intended for customers to use (at least as
+mocks), they are part of the public API.
+
+## The `*Stub` classes
+
+The `*Stub` classes wrap the `*Stub` generated by `gRPC`. They provide several
+functions:
+
+* They are pure interfaces, which allow `google-cloud-cpp` **developers** to
+  mock them in our tests.
+* They are not part of the public interface, they live in `${library}_internal`,
+  and therefore give us some flexibility to change things.
+* They return `StatusOr<T>`, simplifying some of the calls and generic
+  programming.
+* They are the classes most amenable for automatic code generation.
+* When using asynchronous APIs, they return
+  `google::cloud::future<StatusOr<T>>`. This is preferred over using a `void*`
+  to match requests and responses.
+
+The `*Stub` classes are typically organized as a (small) stack of
+[Decorators](https://en.wikipedia.org/wiki/Decorator_pattern), which simplifies
+their testing.
+
+| Layer | Description |
+| ----- | ----------- |
+| `*Logging` | Optional `*Stub` decorator that logs each request and response |
+| `*Metadata` | Injects resource metadata headers for routing |
+| `*RoundRobin` | Round-robins over several `*Stub`s, not all libraries have them |
+
+## Exceptions
```

Comment:
It might be better to use a synonym for "exceptions" when talking about a C++ library.

Reply:
Done.